### DISCUSSION TITLE
Simulating MI300X with gem5 and Saving 100k?

### DISCUSSION THREAD

**Discussion by zevorn** [zevorn/blog#32] [blog]
A while ago, someone in the community published a [Zhihu post on simulating MI300X with gem5][1]. Since I was recently verifying AMDGPU floating-point precision, I wanted to compare the floating-point accuracy of gem5's MI300X model with real hardware.

I recommend running gem5 on a server or workstation, because a personal computer may not have enough resources.

The setup mainly follows that post and the official documentation [Full System AMD GPU model][2].

The basic idea is:

1. Use qemu-system-x86_64 to build an image and kernel that include the AMDGPU driver and ROCm environment, along with the required kernel;
2. Use gem5's KVM mode to emulate an x86 system and integrate the MI300X model as a PCIe card;
3. Finally, load the prepared image and kernel in gem5, and you are ready to use it.

For convenience, I used the Docker environment provided by gem5 to build everything.

```bash
docker pull ghcr.io/gem5/ubuntu-24.04_all-dependencies:v24-0
```

Pull the relevant gem5 repositories locally, create the gem5 Docker container, and share the repository directory with the container:

```bash
git clone https://github.com/gem5/gem5.git
git clone https://github.com/gem5/gem5-resources.git gem5/gem5-resources
docker run --name gem5-amdgpu \
--device /dev/kvm \
--volume /home/zevorn/gem5:/gem5 \
--privileged \
-it ghcr.io/gem5/ubuntu-24.04_all-dependencies:v24-0 
```

The next step is the build stage, which has two parts.

First, create the image and kernel in the gem5/gem5-resources repository outside the container:

```bash
cd gem5/gem5-resources/src/x86-ubuntu-gpu-ml/
./build.sh
```

Be patient here. This build script uses packer to create the image. Once the build succeeds, it produces the following files:

```bash
├── disk-image
│   └── x86-ubuntu-gpu-ml
├── vmlinux-gpu-ml
```

Then enter the container and start building gem5:

```bash
cd /gem5
scons build/VEGA_X86/gem5.opt -j$(nproc)
```

We need to modify the default MI300X Python file so it does not exit after execution:

```python
$ git diff configs/example/gpufs/mi300.py
@@ -153,7 +153,7 @@ def runMI300GPUFS(
         )
         b64file.write(runscriptStr)
 
-    args.script = tempRunscript
+    # args.script = tempRunscript
 
     # Defaults for CPU
     args.cpu_type = "X86KvmCPU"
```

In the gem5 source tree, create a `pytorch_test.py` file for gem5 to load when running MI300, although it will not actually be executed:

```python
#!/usr/bin/env python3

import torch

x = torch.rand(5, 3).to('cuda')
y = torch.rand(3, 5).to('cuda')

z = x @ y
```

We also install `tmux` in the container so that, after starting gem5, we can connect to the terminal from another pane:

```bash
apt update
apt install tmux
```

Then try starting it:

```bash
build/VEGA_X86/gem5.opt configs/example/gpufs/mi300.py --disk-image gem5-resources/src/x86-ubuntu-gpu-ml/disk-image/x86-ubuntu-gpu-ml --kernel gem5-resources/src/x86-ubuntu-gpu-ml/vmlinux-gpu-ml --app pytorch_test.py
```

Using `tmux`, create another terminal in the current window, then connect to gem5 with the following command:

```bash
./util/term/gem5term localhost 3456
root@gem5:~#
```

Next, we need to manually load the AMDGPU driver. You can enter the following commands directly in the gem5 terminal:

```bash
export LD_LIBRARY_PATH=/opt/rocm/lib:$LD_LIBRARY_PATH
export HSA_ENABLE_INTERRUPT=0
export HCC_AMDGPU_TARGET=gfx942
dmesg -n8
cat /proc/cpuinfo
dd if=/root/roms/mi300.rom of=/dev/mem bs=1k seek=768 count=128
```

Let's verify that it loaded successfully:

```bash
root@gem5:~# rocminfo
ROCk module version 6.12.12 is loaded
=====================
HSA System Attributes
=====================
Runtime Version:         1.15
Runtime Ext Version:     1.7
System Timestamp Freq.:  0.001000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE
System Endianness:       LITTLE
Mwaitx:                  DISABLED
XNACK enabled:           NO
DMAbuf Support:          YES
VMM Support:             YES
...
*******
Agent 2
*******
  Name:                    gfx942
  Uuid:                    GPU-XX
  Marketing Name:          AMD Instinct MI300X
  Vendor Name:             AMD
  Feature:                 KERNEL_DISPATCH
  Profile:                 BASE_PROFILE
  Float Round Mode:        NEAR
  Max Queue Number:        128(0x80)
  Queue Min Size:          64(0x40)
  Queue Max Size:          131072(0x20000)
  Queue Type:              MULTI
  Node:                    1
  Device Type:             GPU
...
```

At this point, you can use it normally.

When I tested the fma instruction results on gem5 MI300X, I compared them bit-for-bit against the CPU and real MI300X hardware and found they were always off by 3–4 ulps. I then read the fma source and found that gem5 was simulating it with separate multiply and add steps, which introduced floating-point precision errors relative to real hardware, so I changed the source:

```c
$ git diff src/arch/amdgpu/vega/insts/instructions.hh
diff --git a/src/arch/amdgpu/vega/insts/instructions.hh b/src/arch/amdgpu/vega/insts/instructions.hh
index 7a328f9230..76bd96beee 100644
--- a/src/arch/amdgpu/vega/insts/instructions.hh
+++ b/src/arch/amdgpu/vega/insts/instructions.hh
@@ -44352,8 +44352,9 @@ namespace VegaISA
                         int lane_A = i + M * (block + B * (k / K_L));
                         int lane_B = j + N * (block + B * (k / K_L));
                         int item = k % K_L;
-                        result[i][j] +=
-                          src0[item][lane_A] * src1[item][lane_B];
+                        result[i][j] =
+                            std::fma(src0[item][lane_A], src1[item][lane_B],
+                                     result[i][j]);
                     }
                 }
             }
```

After that, the test behaved correctly. I contributed this bug fix upstream, and it has now been merged: [arch-vega: Improve MFMA precision to match MI300X hardware][3].

[1]: https://www.zhihu.com/question/1928454067929194650
[2]: https://www.gem5.org/documentation/general_docs/gpu_models/gpufs
[3]: https://github.com/gem5/gem5/pull/2478
